File: cpp/67_add_binary/add_binary.cpp
#include <cassert>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  string addBinary(string a, string b) {
    string result = "";
    int i = a.length() - 1, j = b.length() - 1;
    int carry = 0;

    while (i >= 0 || j >= 0 || carry) {
      int sum = carry;
      if (i >= 0) {
        sum += a[i--] - '0';
      }
      if (j >= 0) {
        sum += b[j--] - '0';
      }
      result = char(sum % 2 + '0') + result;
      carry = sum / 2;
    }

    return result;
  }
};

int main() {
  vector<tuple<string, string, string>> tests{
      {"11", "1", "100"},
      {"1010", "1011", "10101"},
  };

  for (auto& [a, b, ans] : tests) {
    assert(Solution().addBinary(a, b) == ans);
  }
}

File: golang/67_add_binary/add_binary.go
// Package main ...
package main

import (
	"math/big"
	"testing"

	"github.com/stretchr/testify/assert"
)

func addBinary(a string, b string) string {
	if a == "" {
		return b
	}
	if b == "" {
		return a
	}

	x := new(big.Int)
	x.SetString(a, 2)
	y := new(big.Int)
	y.SetString(b, 2)
	zero := new(big.Int)
	for y.Cmp(zero) != 0 {
		answer := new(big.Int)
		answer.Xor(x, y)

		carry := new(big.Int)
		carry.And(x, y)
		carry.Lsh(carry, 1)

		x.Set(answer)
		y.Set(carry)
	}

	return x.Text(2)
}

func main() {
	tests := []struct {
		a   string
		b   string
		ans string
	}{
		{"11", "1", "100"},
		{"1010", "1011", "10101"},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, addBinary(test.a, test.b), index)
	}
}
